### PR TITLE
Add support for passing nodeSelector & tolerations to Orchestrator

### LINKED ISSUE #3852: [FR] Helm Chart improvements: nodeSelector and tolerations deployments - part 2
This feature request follows #3848 closed prematurely.

It would be very helpful to have the ability to define `nodeSelector` and `tolerations` for the deployments of `seldon-controller-manager` when installed with the chart.

Judging by the template, this is currently not possible: https://github.com/SeldonIO/seldon-core/blob/master/helm-charts/seldon-core-operator/templates/deployment_seldon-controller-manager.yaml

Ref:
* [kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/)
* [kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration
](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)

### PR DESCRIPTION
**Pull request by hjilke**
**What this PR does / why we need it**:

Adding support for passing nodeSelector & tolerations to the deployment of the Seldon orchestrator. Hopefully, this is something useful.

**Which issue(s) this PR fixes**:

Fixes #3852

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **adriangonz**
You can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by axsaucedo** [comment]
Thank you for the contribution @hjilke - the way that we actually create our Helm charts is actually through a process that converts the "source-of-truth" which is our kustomize objects under `operator/config` into the helm charts using a python script under `operator/helm/split_resources.py`. If you wish to add any new fields to the helm chart you will need to do it through that script, as opposed to directly into the file

**Comment by hjilke** [comment]
Ok cool, let me have a look tomorrow!

**Comment by hjilke** [comment]
Hey @axsaucedo,

**Comment by axsaucedo** [comment]
@hjilke yes that sounds reasonable in principle, let us know if you get any issues

**Comment by hjilke** [comment]
Hey @axsaucedo

**Comment by hjilke** [comment]
Any feedback @axsaucedo?

**Comment by hjilke** [comment]
@axsaucedo done! Thanks for the feedback. Let me know, if there are any further changes required

**Comment by axsaucedo** [comment]
@hjilke unfortunately it seems that there are still some issues, but almost there - the issue is that currently the value is being set as a string, as opposed to an array:

image

You can test the local install with:

```
helm upgrade --install seldon-core helm-charts/seldon-core-operator/ --namespace seldon-system # --set istio.enabled="true" --set istio.gateway="seldon-gateway.istio-system.svc.cluster.local"
```

**Comment by hjilke** [comment]
Ok, perfect. I will update the PR tomorrow!

**Comment by hjilke** [comment]
Hey @axsaucedo I could fix the issues with the type of tolerations. However, I still face:

```
 # helm upgrade --install seldon-core helm-charts/seldon-core-operator/ --namespace seldon-system 

Release "seldon-core" does not exist. Installing it now.
Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: ValidationError(Deployment.spec.template.spec.nodeSelector): invalid type for io.k8s.api.core.v1.PodSpec.nodeSelector: got "string", expected "map"
```
I understand the message, but do you have any suggestions how to fix that?

**Comment by axsaucedo** [comment]
@hjilke certainly - this would require removing the quotation marks, although it may require extending the way that the script performs this replace, you could have a look at other examples where the resulting replacements are a map that doesn't get added with strings

**Comment by hjilke** [comment]
@axsaucedo for both I had already quotations marks removed. What seems to work now is parsing the nodeSelector as yaml in _deployment_seldon-controller-manager.yaml_.

Now installing the charts works.

**Comment by hjilke** [comment]
@axsaucedo any feedback?

File: helm-charts/seldon-core-operator/values.yaml
Comment (axsaucedo) [inline helm-charts/seldon-core-operator/values.yaml:92]: This should actually be a list as opposed to a dictionary (https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/)

File: helm-charts/seldon-core-operator/templates/deployment_seldon-controller-manager.yaml
Comment (axsaucedo) [inline helm-charts/seldon-core-operator/templates/deployment_seldon-controller-manager.yaml:163]: It seems this would be passed as empty if not provided - it doesn't seem to error, but it does seem it may be better to ignore the field completely with a guard (similar to the values above)

**Review by axsaucedo — CHANGES_REQUESTED** [review]
Apologies for the delayed reply, thank you for the efforts - I have tested it locally and it seems to work, I added a couple of minor comments, we are keen to get this in the current release for this/next week, let me know if you don't have a chance to make/test these changes and we can do them on our side. Thank you @hjilke

**Comment by hjilke** [comment]
Sorry, I am spending my free time here and review takes more than 23 days. I dont think that is fair. Closing PR now.

**Comment by axsaucedo** [comment]
That's fair @hjilke, apologies as it's been quite busy at the Seldon side - we'll pick up on this PR / branch from this efforts and fix the final requests made above, your efforst are very much appreciated.

**Pull request closed without merging**
